### DISCUSSION TITLE
MRTK 3 World Locking Tools

### DISCUSSION THREAD

**Discussion by XRWebland** [microsoft/MixedRealityToolkit-Unity#11628] [General]
I'm seeing issues integrating the WLT Core package on the Microsoft Mixed Reality Feature Tool into an MRTK 3 project. Has anyone had any success doing this already and have any tips to the errors I'm seeing? 
![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/127354319/e7cb5874-02de-43d9-8fcb-ae06590cbccf)

**Comment by XRWebland** [comment]
for awareness I was able to get this working with MRTK 3 and Unity 2022.2.12f1. This was an issue not with MRTK 3 but with the changes in Unity and I had to learn how to update this older package for that. 
Update the 2 asmdef files in your Worldlocking.Core package
Search for:
Microsoft.MixedReality.WorldLocking.Core
WorldLocking.Tools.Editor

add in the assembly definition reference to both: Unity.XR.CoreUtils

![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/127354319/70cf312c-6867-4e7e-a599-3474672b2305)

**Reply by Venkat-Swaraj** [reply]
Are you able to use WLT now?

**Reply by Sjiun** [reply]
Thanks for sharing. Came back to this post after reinstalling WLT in another project. Can confirm that this change still fixes the aforementioned compilation errors (Unity 2022.3.5f1 with MRTK 2).

**Reply by laultman** [reply]
It does work. It did change the camera position from the original causing me to have to adjust the app's origin on start.

**Reply by pixelblue** [reply]
Do you mind sharing how you changed the assembly? Both of them are imported as packages and i am unable to change them. 
Did you end up having to import the packages into unity directly and not using them as packages?

**Reply by pixelblue** [reply]
For anyone else looking for this. Here is how you can make changes to your packages by moving them to a different folder:
https://support.unity.com/hc/en-us/articles/9113460764052-How-can-I-modify-built-in-packages

**Reply by Sjiun** [reply]
Sorry, just now read your comment. Thanks for sharing your approach - glad it worked! Personally, I went in and changed the file content since the Inspector window was unmodifiable for me as well. Inside the code, I appended the `Unity.XR.CoreUtils` to the list that contains the other entries shown in the image

**Reply by pixelblue** [reply]
Thanks for your reply.
I am still struggling to get this setup unfortunately.  Do you mind sharing your World Locking Context setup? 
Which Anchor Management are you using?
I am on Unity 2022.3.9LTS, ARFoundation 5.1.0, MRTK3.  Attached a screenshot of my setup
![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/5111190/7bf402c6-ee0c-4f74-ac82-830fd95c99be)

**Reply by Sjiun** [reply]
The following is my WorldLockingContext with the differing versions: Unity 2022.3.5f1, MRTK2. I am using WSA for the Anchor subsystem. I must add that: although I have been able to get the sample apps to work with this project, I have yet to figure out how to save and load the anchors. And even though I can perceive the effect of the markers in i.e. the raypins-example moving back from their drifted-off positions to the positions that I had assigned them to in space, it might still be the case that somehow no anchors were created as described in [this answer on SO](https://stackoverflow.com/questions/74702305/unity-world-locking-tools-persistence-not-working-on-hololens2#answer-74741204). But in my usecase, the save-load functionality was not required and the positioning worked well enough so I did not look into it further.

![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/59316908/35b0a9ae-2094-4919-98dc-6e5247f079cd)

**Reply by pixelblue** [reply]
I see. You are still working with MRTK2. 
Getting WLT working with MRTK3 seems to be a hole other beast to tackle. ;(
Someone did a WLT port for MRTK3. Trying it out now:
https://github.com/Fuzzy-Door-Tech/WLT-MRTK3-Port

**Comment by laultman** [comment]
This fix does allow the code to compile successfully. To get the fix into the code required a bit of effort. If using the Mixed Reality Tool Kit app to add the WLT to your project, it will cause compile-time errors. Unity offers "safe mode" as the recommended option. That is a dead-end. Instead select "Ignore" and allow Unity to import with errors. This allows Unity to complete the Visual Studio project/solution. In Unity follow the instructions given above adding the missing references to the files as directed.

After adding the references you will still need to enable the WLT via the Mixed Reality top menu item. In my case the camera adjustments WLT makes reversed the camera such that the scene was behind me.

**Comment by pixelblue** [comment]
Can someone please share how they got MRTK3 WLT working? What's your exact configuration? 
I managed to get rid of the errors described above but i get warnings that my Anchoring subsystem is not included. I tried WSA, XRSDK, ARFoundation anchoring and etc. but without luck.
Using Unity 2022.3.9LTS, ARFoundation 5.1.0 and MRTK3.  Highly appreciated.

**Reply by laultman** [reply]
@pixelblue Here is my test project nothing happening here, just a cube and the WLT added to the project as I described above. Unity 2022.3.9f1.
I linked it to my [OneDrive](https://1drv.ms/u/s!AmRR17CPJTVK5RGfOr4pNQRHMqC6?e=jkiXPm) as it is over 450mb. The project is set to use holographic remoting and I have it mapped to my HoloLens 2 IP address. You likely will need to adjust for your local network. I have not corrected or changed any settings. Mostly right out of the box. The link is only good for 7 days.

**Reply by pixelblue** [reply]
Highly appreciated it. Tried it out but i get a bunch of errors when trying it in Unity editor. You don't get those errors?:
![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/5111190/dfd6251c-7d75-468f-9d9c-b76978e92f0c)

**Reply by coltonix** [reply]
Hi @pixelblue have you resolved the errors successfully? I'm working with MRTK3 and Unity 2021.3.30LTS so hopefully whatever solved your issues can be helpful to my situation.

The WLT port for MRTK3 that you shared above.. is it working for you?
https://github.com/Fuzzy-Door-Tech/WLT-MRTK3-Port

**Reply by pixelblue** [reply]
I gave up on it and switched back to MRTK2.8. 
From my research WLT doesn't seem to be maintained anymore and i wasn't really able to get it working successfully.

**Reply by coltonix** [reply]
@pixelblue Oh that is unfortunate, can I ask what you used in place of WLT then?

**Reply by pixelblue** [reply]
No real replacement for us. In our case we are required to do offline World alignment. If you can do online alignment then you could use "Azure Spatial Anchors". In our case we use QR code alignment. So the user has to look at the QR code and the world and camera get's aligned to that.

**Reply by coltonix** [reply]
@pixelblue In my case I'm actually trying to do offline alignment as well. 

So can I assume that, by doing what you proposed, I can place models enormous in size without it drifting off over time or any other displacement issues? 

Would you also be able to elaborate a bit on how you would do this "space anchoring" using QR code with MRTK3? I only managed to find solutions with MRTK2

Thank you!

**Reply by XRWebland** [reply]
Just for reference on this, the original fix I posted was to get it working with MRTK3 specifically so I could do it offline.
